### PR TITLE
Add functor material property shortcut output

### PR DESCRIPTION
**Pull request by GiudGiud**
refs #19382 

this will shave the VTB near a hundred lines :)

### CODE REVIEW

**Comment by GiudGiud** [comment]
Patch for PGH is ready

**Comment by lindsayad** [comment]
Is this NEAMS MP TA work?

**Comment by GiudGiud** [comment]
not really I didnt charge anyone for this

File: framework/src/actions/MaterialOutputAction.C [hunk 1]
```diff
@@ -152,30 +162,36 @@ MaterialOutputAction::act()
...
-      // or if the list is empty (all properties)
-      const std::set<std::string> names = mat->getSuppliedItems();
+      // Get all material properties supplied by this material as a starting point
+      std::set<std::string> names = mat->getSuppliedItems();
+      if (dynamic_cast<const FunctorMaterial *>(mat.get()))
+      {
+        const auto fmat_ptr = dynamic_cast<const FunctorMaterial *>(mat.get());
+        names.insert(fmat_ptr->getSuppliedFunctors().begin(),
+                     fmat_ptr->getSuppliedFunctors().end());
+      }
```
Comment (lindsayad) [inline framework/src/actions/MaterialOutputAction.C:172]:
```suggestion
      if (const auto fmat_ptr = dynamic_cast<const FunctorMaterial *>(mat.get()))
        names.insert(fmat_ptr->getSuppliedFunctors().begin(),
                     fmat_ptr->getSuppliedFunctors().end());
```

File: framework/src/actions/MaterialOutputAction.C [hunk 2]
```diff
@@ -491,3 +524,45 @@ MaterialOutputAction::materialOutputHelper<RankFourTensor>(const std::string & p
...
+
+    if (!get_names_only)
+    {
+      auto params = getParams("FunctorVectorElementalAux", property_name, oss.str(), material);
+      params.set<unsigned int>("component") = i;
+      _problem->addAuxKernel("FunctorVectorElementalAux", material.name() + oss.str(), params);
+    }
+  }
+
+  return names;
```
Comment (lindsayad) [inline framework/src/actions/MaterialOutputAction.C:567]: It seems like the only difference between this and the regular material property routine is the aux kernel type name. Can we factor out a helper that takes the aux kernel type name and re-use the code?

File: test/tests/functormaterials/output/tests [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+      detail = 'outputting a functor property, defined within a functor material;'
+    []
+    [ad]
+      # Test the ability for Output objects to enable all material property output
+      type = 'Exodiff'
+      input = 'output_ad.i'
+      exodiff = 'output_ad_out.e'
+
+      detail = 'outputting a functor property with automatic differentiation, defined within a functor material;'
```
Comment (lindsayad) [inline test/tests/functormaterials/output/tests:21]:
```suggestion
      detail = 'outputting a functor property with automatic differentiation, defined within a functor material.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/actions/MaterialOutputAction.C (hunk 2 above)
Comment (GiudGiud) [reply framework/src/actions/MaterialOutputAction.C:567]: there s the addition of the out suffix too and the lack of switch between AD and non AD.

I can go either way

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
it's a nice addition!

**Pull request merged**